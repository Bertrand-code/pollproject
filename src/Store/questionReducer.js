import { fetchLeaderBoardStat } from "../db";
import { _saveQuestion, _saveQuestionAnswer } from "../db/_DATA";

const initialState = {
  questions: [],
  answers: {},
  leaderboard: null,
};

export default function questionReducer(
  state = initialState,
  { type, payload }
) {
  switch (type) {
    case "question_loaded":
      return {
        ...state,
        questions: [...payload.questions],
        answers: { ...state.answers, ...payload.answers },
      };
    case "answer_saved":
      return {
        ...state,
        answers: { ...state.answers, [payload.qid]: payload.answer },
      };
    case "question_saved":
      return {
        ...state,
        questions: [...state.questions, payload],
      };
    case "leaderboard_loaded":
      return {
        ...state,
        leaderboard: payload,
      };

    case "logout":
      return { questions: [], answers: {}, leaderboard: null };

    default:
      return state;
  }
}

export const loadQuestionAction = (questions, answers) => {
  return {
    type: "question_loaded",
    payload: {
      questions,
      answers,
    },
  };
};

export const saveAnswerAction = (qid, answer) => {
  return {
    type: "answer_saved",
    payload: {
      qid,
      answer,
    },
  };
};
const loadLeaderboardAction = (leaderboard) => {
  return {
    type: "leaderboard_loaded",
    payload: leaderboard,
  };
};

const questionSaved = (payload) => {
  return {
    type: "question_saved",
    payload,
  };
};

export const saveQuestionAction = (question) => {
  return async (dispatch, getState) => {
    const { id: userId } = getState().authStore.user;
    const { id } = await _saveQuestion({
      author: userId,
      optionOneText: question.optionOne,
      optionTwoText: question.optionTwo,
    });
    dispatch(questionSaved(id));
  };
};

export const answerQuestion = (qid, answer) => {
  return async (dispatch, getState) => {
    const userId = getState().authStore.user.id;
    _saveQuestionAnswer({ authedUser: userId, qid, answer }).then((res) => {
      if (res) dispatch(saveAnswerAction(qid, answer));
    });
  };
};

export const loadLeaderboard = () => {
  return async (dispatch) => {
    const leaderboard = await fetchLeaderBoardStat();
    dispatch(loadLeaderboardAction(leaderboard));
  };
};
