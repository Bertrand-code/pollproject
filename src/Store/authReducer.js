import { login } from "../db";
import { loadQuestionAction } from "./questionReducer";

const initialValues = {
  user: null,
  auth: false,
  loading: false,
  error: null,
};

export default function authReducer(state = initialValues, { type, payload }) {
  switch (type) {
    case "user_logged":
      return { ...state, auth: true, user: payload };
    case "loading":
      return { ...state, loading: !state.loading };
    case "error":
      return { ...state, error: payload };
    case "logout":
      return { ...state, auth: false, user: null };
    default:
      return state;
  }
}

const loginAction = (payload) => ({ type: "user_logged", payload });
const loadingAction = () => ({ type: "loading" });
export const logoutAction = () => ({ type: "logout" });
export const userLogin = (userInfo) => async (dispatch) => {
  dispatch(loadingAction());
  try {
    const { answers, questions, ...rest } = await login(
      userInfo.username,
      userInfo.password
    );
    dispatch(loginAction(rest));
    dispatch(loadQuestionAction(questions, answers));
  } catch (error) {
    dispatch(loadingAction());
    dispatch({ type: "error", payload: error.message });
  }
};
