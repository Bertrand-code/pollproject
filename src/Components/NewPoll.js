import React from "react";
import { useState } from "react";
import { saveQuestionAction } from "../Store/questionReducer";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { fetchLeaderBoardStat } from "../db";

function NewPoll({ saveQuestion }) {
  const navigate = useNavigate();
  const [question, setQuestion] = useState({
    optionOne: "",
    optionTwo: "",
  });

  const handleChange = (e) => {
    setQuestion({
      ...question,
      [e.target.name]: e.target.value,
    });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    saveQuestion(question);
    navigate("/");
  };

  return (
    <div>
      <form className="newPollForm" onSubmit={handleSubmit}>
        <input
          type="text"
          className="inputElement"
          placeholder="Option One"
          name="optionOne"
          onChange={handleChange}
        />
        <input
          className="inputElement"
          type="text"
          placeholder="Option Two"
          name="optionTwo"
          onChange={handleChange}
        />
        <button className="btn">Add</button>
      </form>
    </div>
  );
}

const mapDispatchToProps = {
  saveQuestion: saveQuestionAction,
  updateLeaderBoard: fetchLeaderBoardStat,
};
export default connect(null, mapDispatchToProps)(NewPoll);
