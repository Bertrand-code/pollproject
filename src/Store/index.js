import {
  applyMiddleware,
  combineReducers,
  compose,
  legacy_createStore as createStore,
} from "redux";
import thunk from "redux-thunk";
import authReducer from "./authReducer";
import questionReducer from "./questionReducer";

export const reducer = combineReducers({
  authStore: authReducer,
  questionsStore: questionReducer,
});

export const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk)
    // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  )
);
