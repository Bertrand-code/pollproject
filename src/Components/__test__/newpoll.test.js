import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import NewPoll from "../NewPoll";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../../Store";

afterEach(() => {
  cleanup();
});

const TestNewPoll = () => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <Routes>
          <Route path="/" element={<NewPoll />} />
        </Routes>
      </Provider>
    </BrowserRouter>
  );
};
describe("newpoll component", () => {
  it("should render and match screenshot", () => {
    const { container } = render(<TestNewPoll />);
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should render form", () => {
    render(<TestNewPoll />);
    expect(screen.getByPlaceholderText("Option One")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Option Two")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /add/i })).toBeInTheDocument();
  });

  it("should user be able to vote", () => {
    render(<TestNewPoll />);
    const optionOne = screen.getByPlaceholderText("Option One");
    const optionTwo = screen.getByPlaceholderText("Option Two");
    fireEvent.click(optionOne);
    fireEvent.change(optionOne, { target: { value: "option One" } });
    fireEvent.click(optionTwo);
    fireEvent.change(optionTwo, { target: { value: "option Two" } });
    expect(optionOne.value).toBe("option One");
    expect(optionTwo.value).toBe("option Two");
  });
});
