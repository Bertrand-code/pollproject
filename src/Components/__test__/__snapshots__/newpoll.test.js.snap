// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newpoll component should render and match screenshot 1`] = `
<div>
  <div>
    <form
      class="newPollForm"
    >
      <input
        class="inputElement"
        name="optionOne"
        placeholder="Option One"
        type="text"
      />
      <input
        class="inputElement"
        name="optionTwo"
        placeholder="Option Two"
        type="text"
      />
      <button
        class="btn"
      >
        Add
      </button>
    </form>
  </div>
</div>
`;
